Do not fetch reserved roles from native store when Get Role API is called

slobodanadamovic opened the pull request:
The reserved roles are already returned directly from the `ReservedRolesStore` in `TransportGetRolesAction`. 
There is no need to query and deserialize reserved roles from the `.security` index just to be merged with static definitions.

slobodanadamovic committed:
Do not fetch reserved roles from native store when Get Role API is called

The reserved roles are already returned from the `ReservedRolesStore`
in `TransportGetRolesAction`. There is no need to query and deserialize
reserved roles from the `.security` index when Get Roles API is called.

slobodanadamovic committed:
Update docs/changelog/121971.yaml

elasticsearchmachine commented:
Hi @slobodanadamovic, I've created a changelog YAML for you.

slobodanadamovic committed:
filter reserved roles in other cases and add tests

slobodanadamovic committed:
Merge branch 'main' into sa-filter-reserved-roles-from-get-roles-api

slobodanadamovic committed:
make roles.yml file updates atomic

slobodanadamovic commented on test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterFactory.java:
```diff
@@ -576,13 +608,11 @@ private void configureSecurity() {
-                                    Files.delete(rolesFile);
-                                    Files.copy(distributionDir.resolve("config").resolve("roles.yml"), rolesFile);
-                                    writeRolesFile();
+                                    updateRolesFileAtomically();
```
This is a fix for Serverless tests.

slobodanadamovic committed:
naming nit

slobodanadamovic committed:
Merge branch 'main' into sa-filter-reserved-roles-from-get-roles-api

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

slobodanadamovic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java:
```diff
@@ -186,15 +199,25 @@ public void getRoleDescriptors(Set<String> names, final ActionListener<RoleRetri
             return;
         }
 
+        final Set<String> rolesToGet = filterReservedRoleNames(names);
```
Filtering all roles here to avoid that reserved roles get resolved by `NativeRolesStore` when `getRoleDescriptors` is called from `accept` method to resolve roles and later used during authorization.

n1v0lg replied:
Hmm I'm not sure that's necessary: 

We [remove](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/RoleDescriptorStore.java#L300-L303) already resolved roles out before handing off to the next resolver in the chain.

So if we are resolving e.g. `viewer` in stateful, it'll get returned by the reserved role store, and won't get checked by the native role store. Similarly, the file store will come first so any role that gets resolved there won't get resolved by the native role store.

We could make this an assertion here (i.e., no reserved roles ever make it to this point).

slobodanadamovic replied:
Nice catch! I've misted the fact that resolved are removed. It makes sense to change it to an assertion.

slobodanadamovic replied:
The filtering is still needed for Serverless, which was the main reason I've added it.
I've introduced it back in the [`TransportGetRolesAction`](https://github.com/elastic/elasticsearch/pull/121971/files#diff-b3d9b568891fc844f44739495b9a3551f100a0d8d99ebd1222ed5ffdcab59b95R56-R61). I think this is okay given that we already do filtering of reserved roles there. The main change is using the `ReservedRoleNameChecker` instead of `ReservedRolesStore`.

slobodanadamovic replied:
Would you mind taking another look?

n1v0lg replied:
Yikes, sorry about that. You're totally right -- we hit a different path when fetching by name. Adding the check sounds good to me. Taking a quick look now 👀

slobodanadamovic commented on x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/GetRolesIT.java:
```diff
@@ -0,0 +1,166 @@
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.Matchers.equalTo;
+
+public class GetRolesIT extends SecurityInBasicRestTestCase {
```
I've added this integration test mostly for sanity check that `GET _security/roles` works as expected in Stateful and returns reserved roles as well.

n1v0lg reviewed (approved):
LGTM -- one suggestion around redundant code.

slobodanadamovic committed:
remove unnecessary reserved roles filtering and add an asserition

slobodanadamovic committed:
Merge branch 'main' into sa-filter-reserved-roles-from-get-roles-api

slobodanadamovic committed:
filter reserved roles

slobodanadamovic committed:
fix transport get action tests

the native roles store should never be called to resolve reserved built-in
roles

slobodanadamovic committed:
use ReservedRoleNameChecker.isReserved consistently

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-filter-reserved-roles-from-get-roles-api

slobodanadamovic committed:
Merge branch 'main' into sa-filter-reserved-roles-from-get-roles-api

slobodanadamovic committed:
Merge branch 'main' into sa-filter-reserved-roles-from-get-roles-api

slobodanadamovic committed:
Merge branch 'main' into sa-filter-reserved-roles-from-get-roles-api

n1v0lg commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/role/TransportGetRolesAction.java:
```diff
@@ -43,23 +51,25 @@ protected void doExecute(Task task, final GetRolesRequest request, final ActionL
                     if (rd != null) {
                         reservedRoles.add(rd);
-                    } else {
-                        listener.onFailure(new IllegalStateException("unable to obtain reserved role [" + role + "]"));
```
We're getting rid of this branch because now `reservedRoleNameChecker.isReserved(...)` can also match on file-based roles in Serverless, right?

n1v0lg reviewed (approved):
LGTM!

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.18](https://github.com/elastic/elasticsearch/pull/122752)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/122752">](https://github.com/elastic/elasticsearch/pull/122752) |
| ✅ |  [9.0](https://github.com/elastic/elasticsearch/pull/122753)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/122753">](https://github.com/elastic/elasticsearch/pull/122753) |
